### DISCUSSION TITLE
Update MOUL banner and StartUp Age background occasionally

### DISCUSSION THREAD

**Discussion by alingo820** [H-uru/Plasma#1388] [Ideas & Proposals]
It would be nice if we could occasionally update the MOUL banner and StartUp Age background that has been showing Teledahn continuously for nearly 20 years now. Seeing those periodically change would make things feel fresher and more alive.

Since it appears that OpenUru/MOULa has started doing major quarter releases (with the 2023Q1 update having recently launched), those would be a good time to do refreshes.

The banner is seen in several places:

- The patcher
- The launcher
- The crash reporter
- The StartUp Age GUI

Many of those have a copy of the same `banner.bmp` that could be replaced, and new builds of those utilities would feature the new artwork. I was not able to pin down what PRP has the StartUp Age GUI that lists all of the player's characters, but that is likely a mipmap in a PRP.

The StartUp Age background is a mipmap in `GUI_District_GUIDialog04a.prp` -- and replacing that with a 16:9 JPEG replaces the background on login.

I mocked up some examples of what new key art would look like, each focusing on a different Age:
# Gahreesen
![StartUp_Gahreesen](https://user-images.githubusercontent.com/68447759/235400129-10d1edbe-a7cb-4245-9365-64537ef13fa3.jpg)

# Eder Kemo
![StartUp_Kemo](https://user-images.githubusercontent.com/68447759/235400332-003982ac-0bee-4b5b-9b54-b9f7dcad2af1.jpg)

# Highgarden
![StartUp_Highgarden](https://user-images.githubusercontent.com/68447759/235400465-34dd2c0c-ddcd-4a54-a52a-cc6bb2d2ecba.jpg)

# Descent
![StartUp_Descent](https://user-images.githubusercontent.com/68447759/235400506-d867c94e-af2d-4206-bf5b-2ee4a97604f3.jpg)

# Elonin (featuring a centered vertical MOUL logo)
![StartUp_Elonin_CenterLogo](https://user-images.githubusercontent.com/68447759/235400666-3b6421e1-b69c-4608-99ed-cf69f84bc134.jpg)

**Comment by dpogue** [comment]
I am 95% sure @Deledrius actually implemented this as a proof-of-concept/demo for https://github.com/H-uru/Plasma/pull/1360... so it's definitely feasible.

There have also been ideas thrown around for over a decade to improve the avatar selection screen by actually showing the avatar in the 3D world: https://web.archive.org/web/20180618031918/http://www.guildofwriters.org/tweek/uru-concepts-login-gui/
I think we've determined that to be technically possible, but not all of the user experience questions

**Reply by alingo820** [reply]
Those look interesting.

In my testing, just replacing the currently embedded bitmaps with new ones work with the current code, with some minor issues I noticed:
1. The StartUp Age presumes the background mipmap is 16:9. If I give it a 21:9 image (to attempt to fill my 21:9 screen all the way), it just stretches the image back to 16:9. I was unable to determine how that ratio is determined/enforced -- via some PRP property or Python code perhaps?
2. The various binaries that embed `banner.bmp` didn't like some of my test images -- they wouldn't render/load in the compiled executables. I think it had to do with some bit depth stuff, but need to do more exploratory testing.

All that said, replacing/updating the StartUp Age background with Python would also be nice, and would allow multiple images that could appear.

I like the idea of featuring background images of new ages/content with each release.

**Reply by dpogue** [reply]
> 1. The StartUp Age presumes the background mipmap is 16:9. If I give it a 21:9 image (to attempt to fill my 21:9 screen all the way), it just stretches the image back to 16:9. I was unable to determine how that ratio is determined/enforced -- via some PRP property or Python code perhaps?

The background is a 3D mesh of a fixed size, and the texture will be applied according to the UV coordinates of that 3D mesh. I'm guessing that mesh is built at a 16:9 ratio right now.

**Reply by alingo820** [reply]
That does seem to be it, it's a 16:9 mesh:
![image](https://user-images.githubusercontent.com/68447759/235407055-4a8c19ff-7e3a-4401-bade-b9dd6919fbc4.png)

Theoretically, I should be able to make a 21:9 plane in Blender. I'll dig into Korman/PyPRP to see if I can get one of those to produce a Plasma Object from the blender mesh I can then import to replace the 16:9 mesh with.

If I can get the mesh replaced, I think it should just work -- the existing 4:3 view of StartUp is the 16:9 mesh, with the sides cut off. I think then that if I provide a native 21:9 image/mesh, then the 16:9 and 4:3 views of StartUp should just have the sides of the 21:9 view cut off. We would just need to make sure we author the new backgrounds to look good in all 3 aspect ratios.

**Reply by Deledrius** [reply]
> I am 95% sure @Deledrius actually implemented this as a proof-of-concept/demo for #1360... so it's definitely feasible.

Not only that, it's been on Gehn for a decade now.  The added flexibility in #1360 is just (stale) icing on the cake. :)

**Reply by Deledrius** [reply]
> If I can get the mesh replaced, I think it should just work -- the existing 4:3 view of StartUp is the 16:9 mesh, with the sides cut off. I think then that if I provide a native 21:9 image/mesh, then the 16:9 and 4:3 views of StartUp should just have the sides of the 21:9 view cut off. We would just need to make sure we author the new backgrounds to look good in all 3 aspect ratios.

My templates have safe regions marked for different aspect ratios, but it's been a very long time and I can't recall if I included 21:9 as a target.  I'm not sure when we adjusted the camera aspect ratio code.  I'd have to double-check!

**Reply by alingo820** [reply]
Some of us were chatting on the OpenUru discord yesterday, and we we think that a target aspect ratio of 32:9 is the  best goal. If we're going the aspect ratio issue, might as well target the common aspect ratios commercial available today.

**Reply by Deledrius** [reply]
The only issue with such a wild aspect ratio is that it does become difficult to compose an image that looks good at so many wildly different widths.  It might make more sense to restrict certain images to certain resolutions, but then we risk bloating the PRP by quite a bit.

**Reply by alingo820** [reply]
Yeah, it is difficult, but not impossible. One of the features planned for my script is it outputting a version of the composited image with the various common aspect ratio 'safe zones' overlaid (4:3, 5:4, 16:9, 21:9), so that the artist can get an idea of what the image would look like at those ratios without needing to load up Uru and change their resolution.

**Reply by dpogue** [reply]
An alternative that's probably acceptable would be to do some sort of vignette effect at the edges and fade them out to black beyond 16:9 or 21:9. That helps keep the focus on the centre of the image (where the login dialog is) while avoiding the worst of perspective distortion.

**Reply by Deledrius** [reply]
That's probably a good solution for the extreme sizes.

**Comment by DurinMephit** [comment]
In general, I would like to see this idea implemented.

That said, please be conscious that some players may wish to not see images of an Age until they have found that place on their own.  E.g. checking if a user has already been to an Age could be implemented as a condition for displaying that Age as a backdrop for login to ensure a spoiler-free experience for even the newest players who have never been to Ages that are like old friends to the rest of us.

if that is too complicated, I suggest at least not "spoiling" new content until at least a quarter after its release.  I have already observed a case where someone advised others to skip the new introductory video in order to avoid "spoilers."

**Reply by dpogue** [reply]
We can't detect player progress on the startup screen because we haven't loaded a player yet. Concern about spoilers is valid though, and even in the core set of Ages we probably want to avoid showing anything too far outside of basic scenery (i.e., the Gahreesen wall might be too much, but the overhead shot of Kemo is probably fine)

**Reply by alingo820** [reply]
Those are valid concerns. But also, I think we should think of new backgrounds as sort of advertisements -- we want to entice returning players into exploration. The game already spoils brand new players today by showing that slice of Teledahn -- and it (at least to me back in '03) was quite enticing when I loaded up Uru: Ages Beyond Myst for the first time and hit that StartUp screen.

Perhaps there are ways to showcase new ages without spoiling too much of them, such as by having the backgrounds avoid any core areas/mechanics of an age, and try to focus on minutia in them. We could ask Age authors if they'd like to submit screenshots to use as background.

In any case, yeah, we would want to come up with general guidelines that try to hit a sweet spot that is neither boring (we only show ages that are old enough to vote) nor too spoiler heavy (here's the solution to the final Kalamee puzzle, happy solving!).

**Reply by Deledrius** [reply]
I think it's generally possible to show Ages from angles and perspectives that are atmospheric and enticing without giving away too much (I've composed several shots along precisely those lines for this purpose).  It requires some mindfulness and knowledge of the Age, and keeping in mind the composition has a large UI Dialog right in the center of the screen. ;)

**Comment by patmauro** [comment]
Re: the spoiler concern - perhaps if the scope of this was limited to the four Journey ages, Ae'Gura, and maybe the Cleft? I think that's plenty to provide the sense of variety this concept is aiming for. Also worth noting fan age content may be specific to certain shards, where Huru in general is trying to provide a baseline - so it might make more sense to start with a "baseline" collection first. Possibly this could be expanded out on a shard-by-shard basis.

By the way - while working on the HD upscale of the launcher banner, I noticed this is actually an entirely _different_ Teledahn banner than the one used in the login screen (note how there's no avatar in the login screen's version). So patcher/launcher might be a separate thing from this (although it'd be cool to apply the idea there as well).

**Reply by alingo820** [reply]
> Re: the spoiler concern - perhaps if the scope of this was limited to the four Journey ages, Ae'Gura, and maybe the Cleft? I think that's plenty to provide the sense of variety this concept is aiming for. Also worth noting fan age content may be specific to certain shards, where Huru in general is trying to provide a baseline - so it might make more sense to start with a "baseline" collection first. Possibly this could be expanded out on a shard-by-shard basis.

A baseline collection makes sense. I'm wondering if it is technically possible to ship a new PRP file containing a set of StartUp Age backgrounds that could be customized by Shard owners to their liking?

Something like a 'branding.prp', that contains:
1.Multiple background images
2. Some sort of map of background images and a weight value that affects their chance of appearing at startup. Something like:
```
kalamee_spoiler_free_bg: 75 //75% chance of showing the newly released (spoiler free) Kalamee age BG
gahreesen_bg: 6 //6% chance of showing the Gahreesen BG
eder_kemo_bg: 6 //6% chance of showing the Eder Kemo BG
eder_gira_bg: 6 //6% chance of showing the Eder Gira BG
kadish_bg: 6 //6% chance of showing the Kadish Tolsea BG
teledahn_classic_bg: 1  //1% chance of showing the classic BG
```
 
If that is doable, then I would say:
1. We should ship a generic version containing only Cyan-created ages.
2. Shard owners can customize `branding.prp` to their heart's content. To build excitement around Shard updates, I would love to see shards customize it around new content releases -- but that's a decision left to shard owners.

I've dug around in the main code branch (and the unmerged stuff in #1360) in an attempt to understand it, but haven't gotten very far. If the above is doable in Python once #1360 is merged, I'd be willing to write it up with some hints/help.

**Comment by Deledrius** [comment]
Sorry for the delay in responding.  Migraines are not fun.

It was a strange sort of déjà vu waking up to a PR for a feature I know I personally added [ten years ago](https://github.com/GehnShard/Plasma/commit/cb7f28082b2e71cfec4144450c65a2e3ec04692f).  We were also tired of staring at the same old ugly Teledahn image back then, so on Gehn we updated the startup background and launcher banner with higher-quality images.  This was long before there were any other active servers (including MOULa), or any asset repositories, unfortunately.  I have the background templates and clean plates made for the various elements (Cyan logo, URU logo, etc.) as well as background images selected for exactly this purpose.  A second update was to add the ability for the game to swap through them, but due to a few setbacks that code has lingered since around 2015 until I finally finished it and submitted it as #1360.  Showcasing (spoiler-free/spoiler-lite) images of the fan-ages on Gehn was precisely the thinking behind adding this feature way back in 2013.

The existing changes on Gehn for this would be the way to go if you want to improve MOULa, as those changes are well-tested and do _exactly_ what you're asking for without you needing to create more from scratch.  The premade elements I have are just waiting for #1360 to be accepted and merged.

I will definitely add the templates and clean elements to the assets repository along with the PR to add the new images to the PRP using an ImageLibModifier.

**Reply by alingo820** [reply]
Sounds like great minds think alike 😉 .

Let's see what we all have, and what the best version of this feature can be. On my side, I've been working on a Python script that will take the various image components (background image, Cyan logo, MOUL logo) and composite them together. That way Shard owners can quickly generate new images without having to fiddle in GIMP or some other image editor.

**Reply by Deledrius** [reply]
Sounds like the script we've used to generate new clothing.  :D

**Reply by alingo820** [reply]
> Sounds like the script we've used to generate new clothing. :D

I was actually looking at that script last night as reference for the Plasma aspect of things.

**Reply by alingo820** [reply]
> I will definitely add the templates and clean elements to the assets repository along with the PR to add the new images to the PRP using an ImageLibModifier.

Any chance you could pass along those clean logo elements via Discord? I can integrate them into my script now.

**Comment by patmauro** [comment]
@DoTshavahn, you may find this useful - @Deledrius was able to point us in the direction of a hi-res "old" Cyan Worlds logo straight from their presskit: https://uru.zero-factorial.com/images/logos/CyanWorldsLogoWLayers.psd

**Reply by alingo820** [reply]
> @DoTshavahn, you may find this useful - @Deledrius was able to point us in the direction of a hi-res "old" Cyan Worlds logo straight from their presskit: https://uru.zero-factorial.com/images/logos/CyanWorldsLogoWLayers.psd

Thanks, that'll do!

**Comment by alingo820** [comment]
Screenshot output by the diagnostic mode of my StartUp background generation script:
![out](https://user-images.githubusercontent.com/68447759/236578664-08da421c-c594-455d-a54d-c2ae15b44557.jpg)

**Reply by alingo820** [reply]
Here's my repo containing the Python script: https://github.com/DoTshavahn/startup-background-generator
